### PR TITLE
Add type definitions for @carbon/icons

### PR DESCRIPTION
**Pull request by metonym**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/carbon__icons/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// Type definitions for @carbon/icons 10.29
+// Project: https://github.com/carbon-design-system/carbon/tree/master/packages/icons
+// Definitions by: Eric Liu <https://github.com/metonym>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 4.0
+
+declare module "@carbon/icons" {
+    type CarbonIconSize = 16 | 20 | 24 | 32;
```
Comment (andrewbranch) [inline types/carbon__icons/index.d.ts:8]: I believe you can unwrap these types and just export them. Does that not work?

Comment (metonym) [reply types/carbon__icons/index.d.ts:8]: I tried that initially but received an error when using it in the `declare module "@carbon/icons/metadata.json"` ambient module.

Comment (andrewbranch) [reply types/carbon__icons/index.d.ts:8]: You don’t need to import them at that point because they’ll be in scope.

Comment (andrewbranch) [reply types/carbon__icons/index.d.ts:8]: Actually, I believe you’ll have to move the declaration for metadata.json to its own file. Then they should play nicely together. (Don’t forget to add `metadata.d.ts` to  `files` in tsconfig.json.)

Comment (metonym) [reply types/carbon__icons/index.d.ts]: @andrewbranch thank you for the suggestion – moving the declaration to a separate file works. The lint/test scripts passed locally but I'm seeing this [error in the CI](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52154/checks?check_run_id=2256184704). Any suggestions?

```
Error: carbon__icons: Only index.d.ts may be listed explicitly in tsconfig's "files" entry.
Other d.ts files must either be referenced through index.d.ts, tests, or added to OTHER_FILES.txt.
```

**Review by andrewbranch — APPROVED** [review]

File: types/carbon__icons/index.d.ts (hunk 1 above)
Comment (metonym) [reply types/carbon__icons/index.d.ts]: Applied the same fix to https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52155

**Comment by metonym** [comment]
@typescript-bot please do not close this. The failing CI is expected

**Pull request closed without merging**